typify-impl/src/merge.rs

```diff
@@ -411,62 +428,83 @@ fn try_merge_schema_not(
+            metadata: _,
+            // TODO we should should check instance_type and then walk through
+            // validation of each type based on the specific validation.
+            instance_type: _,
+            format: _,
+            enum_values: _,
+            const_value: _,
+            subschemas,
+            number: _,
+            string: _,
+            array: _,
+            object,
+            // TODO we might want to chase these references but need to take
+            // care to handle circular references.
+            reference: _,
             extensions: _,
         }) => {
-            // TODO this is incomplete, but seems sufficient for the schemas
-            // we've seen in the wild.
-            if let Some(ObjectValidation {
-                required,
-                properties,
-                ..
-            }) = schema_object.object.as_deref_mut()
-            {
-                not_object.required.iter().for_each(|not_required| {
-                    let _ = required.remove(not_required);
-                    let _ = properties.remove(not_required);
-                });
+            if let Some(not_object) = object {
+                // TODO this is incomplete, but seems sufficient for the
+                // schemas we've seen in the wild.
+                if let Some(ObjectValidation {
+                    required,
+                    properties,
+                    ..
+                }) = schema_object.object.as_deref_mut()
+                {
+                    // TODO This is completely wrong for arrays of len > 1.
+                    // We need to treat required: [x, y] like it's:
+                    //   not:
+                    //     allOf:
+                    //       required: [x]
+                    //       required: [y]
+                    // Then we can transform them into:
+                    //   anyOf:
+                    //     not:
+                    //       required: [x]
+                    //     not:
+                    //       required: [y]
+                    // Which in turn can become:
+                    //   oneOf:
+                    //     not:
+                    //       required: [x]
+                    //     not:
+                    //       required: [y]
+                    //     not:
+                    //       required: [x, y]
+                    for not_required in &not_object.required {
+                        // A property can't be both required and not required
+                        // therefore this schema is unsatisfiable.
+                        if required.contains(not_required) {
+                            return Err(());
+                        }
+                        // We don't care if a property we're saying is not
+                        // required was not present in the properties map.
+                        let _ = properties.remove(not_required);
```

Comment:
alternatively we could do something like this:

```rust
let _ = properties.insert(not_required, Schema::Bool(false));
```

To indicate that this must not be present. The resulting types might be ugly or we could look for `Option<Never>` and do something different.